push_config.go

```diff
@@ -0,0 +1,183 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package elemental
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// A PushFilter represents an abstract filter for filtering out push notifications. This is now aliased to PushConfig as a
+// result of re-naming the type.
+//
+// Deprecated: use the new name PushConfig instead
+type PushFilter = PushConfig
+
+// A PushConfig represents an abstract filter for filtering out push notifications.
+//
+// The 'IdentityFilters' field is a mapping between a filtered identity and the string representation of an elemental.Filter.
+// A client will supply this attribute if they want fine-grained filtering on the set of identities that they are filtering on.
+// If this attribute has been supplied, the identities passed to 'IdentityFilters' must be a subset of the identities passed to
+// 'Identities'; passing in identities that are not provided in the 'Identities' field will be ignored.
+type PushConfig struct {
+	Identities      map[string][]EventType `msgpack:"identities" json:"identities"`
+	IdentityFilters map[string]string      `msgpack:"filters"    json:"filters"`
+	Params          url.Values             `msgpack:"parameters" json:"parameters"`
+
+	// parsedIdentityFilters holds the parsed `IdentityFilters` to avoid re-parsing the configured filters on each push
+	// event that is using the same config.
+	parsedIdentityFilters map[string]*Filter
+}
+
+// NewPushFilter returns a new PushFilter. NewPushFilter is now aliased to NewPushConfig. This was done for backwards
+// compatibility as a result of the re-naming of PushFilter to PushConfig.
+//
+// Deprecated: use the constructor with the new name, NewPushConfig, instead
+func NewPushFilter() *PushFilter {
+	fmt.Println("DEPRECATED: elemental.NewPushFilter is deprecated, use elemental.NewPushConfig instead")
+
+	return &PushFilter{
+		Identities:            map[string][]EventType{},
+		IdentityFilters:       map[string]string{},
+		parsedIdentityFilters: map[string]*Filter{},
+	}
+}
+
+// NewPushConfig returns a new PushConfig.
+func NewPushConfig() *PushConfig {
+
+	return &PushConfig{
+		Identities:            map[string][]EventType{},
+		IdentityFilters:       map[string]string{},
+		parsedIdentityFilters: map[string]*Filter{},
+	}
+}
+
+// SetParameter sets the values of the parameter with the given key.
+func (pc *PushConfig) SetParameter(key string, values ...string) {
+
+	if pc.Params == nil {
+		pc.Params = url.Values{}
+	}
+
+	pc.Params[key] = values
+}
+
+// Parameters returns a copy of all the parameters.
+func (pc *PushConfig) Parameters() url.Values {
+
+	if pc.Params == nil {
+		return nil
+	}
+
+	out := url.Values{}
```

Comment:
`out := make(url.Values, len(pc.Params))` makes it more efficient

Reply:
donezo